Play scala NoClassDefFoundError: play/mvc/Results$Status

I use the atmosphere-play for a play-scala project,  when I connect the server use websocket, I get some error like:

```java 
java.lang.NoClassDefFoundError: play/mvc/Results$Status
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2658)
	at java.lang.Class.getDeclaredConstructors(Class.java:2007)
	at com.google.inject.spi.InjectionPoint.forConstructorOf(InjectionPoint.java:245)
	at com.google.inject.internal.ConstructorBindingImpl.create(ConstructorBindingImpl.java:99)
	at com.google.inject.internal.InjectorImpl.createUninitializedBinding(InjectorImpl.java:658)
	at com.google.inject.internal.InjectorImpl.createJustInTimeBinding(InjectorImpl.java:882)
	at com.google.inject.internal.InjectorImpl.createJustInTimeBindingRecursive(InjectorImpl.java:805)
	at com.google.inject.internal.InjectorImpl.getJustInTimeBinding(InjectorImpl.java:282)
	at com.google.inject.internal.InjectorImpl.getBindingOrThrow(InjectorImpl.java:214)
Caused by: java.lang.ClassNotFoundException: play.mvc.Results$Status
	at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2658)
	at java.lang.Class.getDeclaredConstructors(Class.java:2007)
	at com.google.inject.spi.InjectionPoint.forConstructorOf(InjectionPoint.java:245)
```

here is the source:
Chat.scala
```scala
@Singleton
@ManagedService(path = "/chat")
class Chat @Inject() extends AtmosphereController {
  private val logger = LoggerFactory.getLogger("circle")

  @Ready
  def onReady(resource: AtmosphereResource) = {
    logger.info("Browser {} connected.", resource.uuid())

  }

  @Disconnect
  def onDisconnected(event: AtmosphereResourceEvent) = {
    if (event.isCancelled) {
      logger.info("Browser {} unexpectedly disconnected", event.getResource.uuid())
    } else if (event.isClosedByClient) {
      logger.info("Browser {} closed the connection", event.getResource.uuid())
    }
  }

  @org.atmosphere.config.service.Message
  def onMessage(message: String) = {
    logger.info("message {}", message)
  }

}
```

GlobalSetting.scala
```scala
@Singleton
class GlobalSetting @Inject()(appLifecycle: ApplicationLifecycle) {
  AtmosphereCoordinator.instance().discover(classOf[CircleController]).ready()

  appLifecycle.addStopHook { () =>
    AtmosphereCoordinator.instance().shutdown()
    Future.successful(())
  }

}
```

build.sbt
```scala
libraryDependencies ++= Seq(
  jdbc,
  cache,
  ws,
  "org.scalatestplus.play" %% "scalatestplus-play" % "1.5.0-RC1" % Test,
  "org.atmosphere" % "atmosphere-play" % "2.2.0",
  "org.atmosphere" % "atmosphere-runtime" % "2.4.3",
  "org.atmosphere" % "atmosphere-guice" % "2.4.2"
)
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,6 @@
 sudo: false
 language: java
 jdk:
-  - oraclejdk7
+  - oraclejdk8
 before_install: 'mvn -version'
 install: 'mvn clean install'

--- a/README.md
+++ b/README.md
@@ -53,9 +53,6 @@ For Play 2.1.x:
      </dependency>
 ```
 
-
-
-
 ## Join the growing community
 
 If you are interested, subscribe to our [mailing list](http://groups.google.com/group/atmosphere-framework) for more info!  We are on irc.freenode.net under #atmosphere-comet

--- a/README.md
+++ b/README.md
@@ -13,6 +13,26 @@ Fork the [samples workspace](https://github.com/Atmosphere/atmosphere-samples/tr
 
 Download [Atmosphere Play!](http://search.maven.org/#search%7Cga%7C1%7Catmosphere-play), use Maven or  [sbt](http://ntcoding.blogspot.ca/2013/09/atmosphere-scala-sbt-akka-step-by-step.html)
 
+For Play 2.5.x+:
+
+```xml
+     <dependency>
+         <groupId>org.atmosphere</groupId>
+         <artifactId>atmosphere-play</artifactId>
+         <version>2.3.0</version>
+     </dependency>
+```
+
+For Play 2.4.x+:
+
+```xml
+     <dependency>
+         <groupId>org.atmosphere</groupId>
+         <artifactId>atmosphere-play</artifactId>
+         <version>2.2.0</version>
+     </dependency>
+```
+
 For Play 2.2.x+:
 
 ```xml
@@ -32,6 +52,9 @@ For Play 2.1.x:
          <version>1.0.0</version>
      </dependency>
 ```
+
+
+
 
 ## Join the growing community
 

--- a/module/src/main/java/org/atmosphere/play/AtmosphereController.java
+++ b/module/src/main/java/org/atmosphere/play/AtmosphereController.java
@@ -20,8 +20,8 @@ import org.atmosphere.cpr.AtmosphereConfig;
 import org.atmosphere.cpr.AtmosphereFramework;
 import org.atmosphere.util.IOUtils;
 import play.mvc.Controller;
+import play.mvc.LegacyWebSocket;
 import play.mvc.Result;
-import play.mvc.WebSocket;
 
 import java.util.Collections;
 import java.util.Map;
@@ -45,7 +45,7 @@ public class AtmosphereController extends Controller {
         }
     }
 
-    public WebSocket<String> webSocket() throws Throwable {
+    public LegacyWebSocket<String> webSocket() throws Throwable {
         return new PlayWebSocket(config, request(), convertedSession()).internal();
     }
 

--- a/module/src/main/java/org/atmosphere/play/AtmosphereUtils.java
+++ b/module/src/main/java/org/atmosphere/play/AtmosphereUtils.java
@@ -65,7 +65,7 @@ public class AtmosphereUtils {
             if (request.body().asText() != null) {
                 body = request.body().asText().getBytes("utf-8");
             } else if (request.body().asRaw() != null) {
-                body = request.body().asRaw().asBytes();
+                body = request.body().asBytes().toArray();
             } else if (request.body().asJson() != null) {
                 body = request.body().asJson().asText().getBytes("utf-8");
                 if (body.length == 0) {

--- a/module/src/main/java/org/atmosphere/play/PlayAsyncIOWriter.java
+++ b/module/src/main/java/org/atmosphere/play/PlayAsyncIOWriter.java
@@ -62,12 +62,7 @@ public class PlayAsyncIOWriter extends AtmosphereInterceptorWriter implements Pl
                 try {
                     final AtmosphereRequest r = AtmosphereUtils.request(request, additionalAttributes);
                     if (transport != null && transport.length > 0 && !transport[0].equalsIgnoreCase(HeaderConfig.POLLING_TRANSPORT)) {
-                        out.onDisconnected(new F.Callback0() {
-                            @Override
-                            public void invoke() throws Throwable {
-                                _close(r);
-                            }
-                        });
+                        out.onDisconnected(() -> _close(r));
                     }
 
                     AtmosphereResponse res = new AtmosphereResponseImpl.Builder()

--- a/module/src/main/java/org/atmosphere/play/PlayWebSocket.java
+++ b/module/src/main/java/org/atmosphere/play/PlayWebSocket.java
@@ -25,18 +25,20 @@ import org.slf4j.LoggerFactory;
 
 import play.libs.F;
 import play.mvc.Http;
+import play.mvc.LegacyWebSocket;
 import play.mvc.WebSocket;
 
 import java.io.IOException;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
 
-public class PlayWebSocket extends org.atmosphere.websocket.WebSocket implements PlayInternal<WebSocket<String>> {
+public class PlayWebSocket extends org.atmosphere.websocket.WebSocket implements PlayInternal<LegacyWebSocket<String>> {
     private static final Logger logger = LoggerFactory.getLogger(PlayWebSocket.class);
     private WebSocket.Out<String> out;
     private WebSocket.In<String> in;
 
-    private final WebSocket<String> w;
+    private final LegacyWebSocket<String> w;
     private final AtomicBoolean firstWrite = new AtomicBoolean(false);
     private final WebSocketProcessor webSocketProcessor;
 
@@ -44,23 +46,13 @@ public class PlayWebSocket extends org.atmosphere.websocket.WebSocket implements
         super(config);
 
         webSocketProcessor = WebSocketProcessorFactory.getDefault().getWebSocketProcessor(config.framework());
-        w = new WebSocket<String>() {
+        w = new LegacyWebSocket<String>() {
             @Override
-            public void onReady(WebSocket.In<String> iin, Out<String> oout) {
+            public void onReady(WebSocket.In<String> iin, WebSocket.Out<String> oout) {
                 out = oout;
                 in = iin;
-                in.onClose(new F.Callback0() {
-                    @Override
-                    public void invoke() throws Throwable {
-                        webSocketProcessor.close(PlayWebSocket.this, 1002);
-                    }
-                });
-                in.onMessage(new F.Callback<String>() {
-                    @Override
-                    public void invoke(String message) throws Throwable {
-                        webSocketProcessor.invokeWebSocketProtocol(PlayWebSocket.this, message);
-                    }
-                });
+                in.onClose(() -> webSocketProcessor.close(PlayWebSocket.this, 1002));
+                in.onMessage(message -> webSocketProcessor.invokeWebSocketProtocol(PlayWebSocket.this, message));
                 AtmosphereRequest r = null;
                 try {
                     r = AtmosphereUtils.request(request, additionalAttributes);
@@ -78,7 +70,7 @@ public class PlayWebSocket extends org.atmosphere.websocket.WebSocket implements
         };
     }
 
-    public WebSocket<String> internal() {
+    public LegacyWebSocket<String> internal() {
         return w;
     }
 

--- a/module/src/main/scala/org/atmosphere/play/AtmosphereHttpRequestHandler.scala
+++ b/module/src/main/scala/org/atmosphere/play/AtmosphereHttpRequestHandler.scala
@@ -1,18 +1,23 @@
 package org.atmosphere.play
 
+import java.util.concurrent.CompletableFuture
+
 import com.google.inject.Inject
 import play.api.http._
+import play.api.inject.Injector
 import play.api.mvc.{Handler, RequestHeader}
 import play.api.routing.Router
 import play.core.j._
-import play.libs.F.Promise
+import play.api.mvc.BodyParser
+import play.mvc.Http.RequestBody
 
 
 class AtmosphereHttpRequestHandler @Inject()(components: JavaHandlerComponents,
                                              router: Router,
                                              errorHandler: HttpErrorHandler,
                                              configuration: HttpConfiguration,
-                                             filters: HttpFilters)
+                                             filters: HttpFilters,
+                                             injector: Injector)
   extends JavaCompatibleHttpRequestHandler(router, errorHandler, configuration, filters, components) {
 
   override def routeRequest(request: RequestHeader) = {
@@ -37,8 +42,7 @@ class AtmosphereHttpRequestHandler @Inject()(components: JavaHandlerComponents,
 
     else {
       val controller: AtmosphereController =
-        Option(play.api.Play.current.injector.instanceOf(controllerClass))
-          .getOrElse(controllerClass.newInstance)
+        Option(injector.instanceOf(controllerClass)).getOrElse(controllerClass.newInstance)
 
       // Netty fail to decode headers separated by a ','
       val javaAction =
@@ -47,15 +51,26 @@ class AtmosphereHttpRequestHandler @Inject()(components: JavaHandlerComponents,
         else
           new JavaAction(components) {
             val annotations = new JavaActionAnnotations(controllerClass, controllerClass.getMethod("http"))
-            val parser = annotations.parser
+            val parser = javaBodyParserToScala(injector.instanceOf(annotations.parser))
 
-            def invocation = Promise.pure(controller.http)
+            def invocation = CompletableFuture.completedFuture(controller.http)
           }
 
       Some(javaAction)
     }
   }
 
+  def javaBodyParserToScala(parser: play.mvc.BodyParser[_]): BodyParser[RequestBody] = BodyParser { request =>
+    val accumulator = parser.apply(new play.core.j.RequestHeaderImpl(request)).asScala()
+    import play.api.libs.iteratee.Execution.Implicits.trampoline
+    accumulator.map { javaEither =>
+      if (javaEither.left.isPresent) {
+        Left(javaEither.left.get().asScala())
+      } else {
+        Right(new RequestBody(javaEither.right.get()))
+      }
+    }
+  }
 
   def dispatch(request: RequestHeader): Option[Handler] =
     dispatch(request, classOf[AtmosphereController])

--- a/pom.xml
+++ b/pom.xml
@@ -47,7 +47,7 @@
             <dependency>
                 <groupId>com.typesafe.play</groupId>
                 <artifactId>play_2.11</artifactId>
-                <version>2.4.6</version>
+                <version>2.5.4</version>
                 <scope>provided</scope>
             </dependency>
             <dependency>
@@ -188,7 +188,7 @@
                 <version>2.7</version>
                 <configuration>
                     <aggregate>true</aggregate>
-                    <source>1.6</source>
+                    <source>1.8</source>
                     <encoding>UTF-8</encoding>
                     <maxmemory>1g</maxmemory>
                     <links>

